packages/core/core-flows/src/draft-order/workflows/update-draft-order.ts

```diff
@@ -0,0 +1,258 @@
+import { Modules, OrderWorkflowEvents } from "@medusajs/framework/utils"
+import {
+  createStep,
+  createWorkflow,
+  StepResponse,
+  transform,
+  WorkflowData,
+  WorkflowResponse,
+} from "@medusajs/framework/workflows-sdk"
+import {
+  IOrderModuleService,
+  OrderDTO,
+  RegisterOrderChangeDTO,
+  UpdateOrderDTO,
+  UpsertOrderAddressDTO,
+} from "@medusajs/types"
+import { emitEventStep, useRemoteQueryStep } from "../../common"
+import { previewOrderChangeStep, registerOrderChangesStep } from "../../order"
+import { validateDraftOrderStep } from "../steps/validate-draft-order"
+
+export const updateDraftOrderWorkflowId = "update-draft-order"
+
+export interface UpdateDraftOrderWorkflowInput {
+  /**
+   * The ID of the draft order to update.
+   */
+  id: string
+  /**
+   * The ID of the user updating the draft order.
+   */
+  user_id: string
+  /**
+   * Create or update the shipping address of the draft order.
+   */
+  shipping_address?: UpsertOrderAddressDTO
+  /**
+   * Create or update the billing address of the draft order.
+   */
+  billing_address?: UpsertOrderAddressDTO
+  /**
+   * The ID of the customer to associate the draft order with.
+   */
+  customer_id?: string
+  /**
+   * The new email of the draft order.
+   */
+  email?: string
+  /**
+   * The ID of the sales channel to associate the draft order with.
+   */
+  sales_channel_id?: string
+  /**
+   * The new metadata of the draft order.
+   */
+  metadata?: Record<string, unknown> | null
+}
+
+interface UpdateDraftOrderStepInput {
+  order: OrderDTO
+  input: UpdateOrderDTO
+}
+
+const updateDraftOrderStep = createStep(
+  "update-draft-order",
+  async ({ order, input }: UpdateDraftOrderStepInput, { container }) => {
+    const service = container.resolve<IOrderModuleService>(Modules.ORDER)
+
+    const updatedOrders = await service.updateOrders([
+      {
+        id: order.id,
+        ...input,
+      },
+    ])
+
+    const updatedOrder = updatedOrders[0]
+
+    return new StepResponse(updatedOrder, order)
+  },
+  async function (prevData, { container }) {
+    if (!prevData) {
+      return
+    }
+
+    const service = container.resolve<IOrderModuleService>(Modules.ORDER)
+
+    await service.updateOrders([prevData as UpdateOrderDTO])
```

Comment:
todo: This doesn't seem right, `prevData` is the new data, so the compensation action would trigger an update with the new data.